### ISSUE TITLE
Json schema $ref no folow link shown for external definition

### ISSUE THREAD

**Issue by ionut-gheorghe** [microsoft/vscode#116540]
- VS Code Version: 1.53.2
- OS Version: Windows_NT x64 10.10.19041

![image](https://user-images.githubusercontent.com/54311614/107764095-e5ba1480-6d2f-11eb-92f8-067d0420e288.png)
![image](https://user-images.githubusercontent.com/54311614/107764409-78f34a00-6d30-11eb-9a63-35efe1cadbec.png)
![image](https://user-images.githubusercontent.com/54311614/107764032-ca4f0980-6d2f-11eb-94f9-a27ea37c697c.png)

 I remember in older version that it showed the link.

Does this issue occur when all extensions are disabled?: Yes

**Comment by aeschli** [comment]
We currently only support the link for local $ref's `#...`.

**Comment by yarally** [comment]
Hello, is this issue currently being worked on? Otherwise I would like to check it out.

**Comment by aeschli** [comment]
I'm not working on it, if you want to give it a go, that would be fantastic.

**Comment by yarally** [comment]
Hi, after investigating for a bit I believe this feature should be added in the JSON language service. Am I correct in assuming that that we need to commit and make a PR for https://github.com/microsoft/vscode-json-languageservice? If so, should this issue be moved to there as well?

**Comment by FilipePintoReis** [comment]
@yarally how did you come to this conclusion?

**Comment by yarally** [comment]
@FilipePintoReis For one of my master courses called Software Architecture we are required to learn about and contribute to a open source project. First of all I noticed that turning off the JSON language service makes it so that the support links stop functioning. Then one of my teammates found [this file](https://github.com/microsoft/vscode-json-languageservice/blob/07fec3faafa3f32915dcd6fcc6999a91ad113f88/src/services/jsonLinks.ts), which we believe handles the links of the $refs.

**Comment by aeschli** [comment]
Yes, the change needs to be in https://github.com/microsoft/vscode-json-languageservice. We can move the issue there if that's helpful, or just create a PR there.

**Comment by yarally** [comment]
Okay thanks! Keeping the issue here is fine.

**Comment by Huntervang** [comment]
Hi I'm a fellow student of @yarally and I am also currently looking into this issue. We think we will be able to implement this by adding a DocumentContext parameter to the [findLinks](https://github.com/microsoft/vscode-json-languageservice/blob/07fec3faafa3f32915dcd6fcc6999a91ad113f88/src/services/jsonLinks.ts#L10) function, similar to how the html-language-service's [link finding function](https://github.com/microsoft/vscode-html-languageservice/blob/2f7ae4df298ac2c299a40e9024d118f4a9dc0c68/src/services/htmlLinks.ts#L83) works. Currently [this](https://github.com/microsoft/vscode/blob/a699ffaee62010c4634d301da2bbdb7646b8d1da/extensions/css-language-features/server/src/utils/documentContext.ts#L11) is how the html and css language services both implement DocumentContext. Would this approach be correct?

**Comment by aeschli** [comment]
@Huntervang What you suggest makes sense!

**Comment by yarally** [comment]
@aeschli We have an almost working proposal at the moment, but there is one problem that we're unsure how to solve. In order to calculate the ranges for the references, we need the respective TextDocuments. Inside the [jsonServer.ts](https://github.com/microsoft/vscode/blob/main/extensions/json-language-features/server/src/jsonServer.ts), this is done using the document manager. However, the manager only has references to documents that are currently open. Is there a way to add documents to the managers's memory or to fetch them some other way using the uri? Thanks in advance!

```ts
if (ref.uri) {
    const externalDocument = documents.get(ref.uri);
    // externalDocument is undefined if the user has not opened the file first
    if (externalDocument) {
    // Do stuff
    }
```

**Comment by aeschli** [comment]
`$ref` points to schemas and for loading schemas we already have the `schemaRequestService`. I suggest to use that to load the references schema content. It will also work correctly with `file` URI's open in an editor.
You can then create a TextDocument and use the jsonc-parser library then will help to parse the content and find line/character of the referenced node.

The schema service uses the schemaRequestService and does caching. However, it it currently only caches the object tree, not the locations nor the underlying document.
I wouldn't add anything to the cache at the moment.

Instead, I would delay the resolving of the $ref location, so we don't need the caching. This can be done with [the document link resolve](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#documentLink_resolve) feature from LSP.
For that, the link provider returns a link without `target`. When the link is actually clicked by the user, the `resolve` call is fired and allows the server to compute a target (and a correct range)
-> new API `resolveLink` on the json language service

**Comment by Huntervang** [comment]
If we were to only resolve the link on click this would result in us not knowing the existence of a target JSON object, and thus not knowing when the link should become a clickable object. We currently have some [code](https://github.com/yarally/vscode-json-languageservice/blob/d65f03df1df221a24ddd2d84575c184396950cf1/src/services/jsonLinks.ts#L30) which checks if the file exists and in that case creates a link. This results in the current functionality as seen below:
![](https://media.discordapp.net/attachments/549685866579230750/829023665085349919/unknown.png)
Here the third link is in fact not a valid link, as the specified JSON object does not exist in the targeted file. 
What would be the desired behavior in such cases?

Another question we had is what would be a function that gets called when the user clicks a link within the scope of [jsonServer.ts](https://github.com/microsoft/vscode/blob/main/extensions/json-language-features/server/src/jsonServer.ts)?

**Comment by aeschli** [comment]
I would always render it as a link, if the link is syntactically correct. Even if it points to a non existing file or property. 
Only when you click you will find out. It's the same with a `href` references in HTLM documents.
Only showing it as link if it really points to something would mean that we have to watch file changes and content changes as these could result in a link to become valid or invalid.

**Comment by aeschli** [comment]
In jsonServer you have to add calls to `connection.onDocumentLinks` and `connection.onDocumentLinkResolve`
